oscarlab/graphene
# [Pal/Linux-SGX] OCALL security hardening

## Affected components


## Description of the changes (reasons and measures)
OCALL security hardening:
  - Time-Of-Check-To-Time-Of-Use (TOCTOU) vulnerability is closed by first copying untrusted values inside enclave and then operating on these copied values; see COPY_FROM_USER
  - Pointer/integer overflow is closed by comparing against a trusted maximum value in COPY_FROM_USER
  - Untrusted stack overflows during OCALLOC/ALLOC_IN_USER/COPY_TO_USER are closed by checking for NULL return values

These vulnerabilities were independently discovered and disclosed by David Oswald, Jo van Bulck, and others.

Fixes #549.

## How to test this PR? (if applicable)

All regression tests should run.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene/622)
